File: sdk/aiep.js
/**
 * AIEP (AI-Agent Identity & Execution Protocol) JavaScript SDK
 * 用于与 AetheriaAgentDID 智能合约交互的官方 SDK。
 * 支持 DID 身份管理、EIP-712 委托支付和执行、以及反事实部署逻辑。
 */

const { ethers } = require("ethers");

/**
 * 获取 EIP-712 域信息
 * @param {number} chainId 链 ID
 * @param {string} verifyingContract 验证签名的合约地址
 */
function getDomain(chainId, verifyingContract) {
  return {
    name: "AetheriaAgentDID",
    version: "1",
    chainId,
    verifyingContract
  };
}

/**
 * EIP-712 结构化数据类型定义
 * 必须与 AetheriaAgentDID.sol 中的 TYPEHASH 定义保持完全一致
 */
const Types = {
  PayEth: [
    { name: "to", type: "address" },
    { name: "amount", type: "uint256" },
    { name: "nonce", type: "uint256" },
    { name: "deadline", type: "uint256" }
  ],
  PayERC20: [
    { name: "token", type: "address" },
    { name: "to", type: "address" },
    { name: "amount", type: "uint256" },
    { name: "nonce", type: "uint256" },
    { name: "deadline", type: "uint256" }
  ],
  Execute: [
    { name: "target", type: "address" },
    { name: "value", type: "uint256" },
    { name: "dataHash", type: "bytes32" },
    { name: "nonce", type: "uint256" },
    { name: "deadline", type: "uint256" }
  ]
};

/**
 * AetheriaAgentDID 核心合约 ABI
 */
const ABI = [
  "function ownerOf() view returns (address)",
  "function setAgentSigner(address)",
  "function transferAgentOwnership(address)",
  "function getNonce() view returns (uint256)",
  "function depositToAgent() payable",
  "function balanceOf() view returns (uint256)",
  "function delegatedPayEth(address,uint256,uint256,bytes)",
  "function depositERC20(address,uint256)",
  "function balanceOfERC20(address) view returns (uint256)",
  "function delegatedPayERC20(address,address,uint256,uint256,bytes)",
  "function updateMetadata(string)",
  "function did() view returns (string)",
  "function isFrozen() view returns (bool)",
  "function getAgentSigner() view returns (address)",
  "function delegatedExecute(address,uint256,bytes,uint256,bytes)"
];

/**
 * AetheriaFactory 工厂合约 ABI
 */
const FACTORY_ABI = [
  "event AgentDeployed(address indexed agent, address indexed owner, address indexed signer)",
  "function deployAgent(address,address,string,bytes32) returns (address)",
  "function computeAddress(address,address,string,bytes32) view returns (address)",
  "function deployAndDelegatedPayERC20(address,address,string,bytes32,address,address,uint256,uint256,bytes) returns (address)",
  "function deployAndDelegatedPayEth(address,address,string,bytes32,address,uint256,uint256,bytes) returns (address)",
  "function deployAndDelegatedExecute(address,address,string,bytes32,address,uint256,bytes,uint256,bytes) returns (address)"
];

/**
 * AetheriaSafeModule 模块 ABI (用于 Gnosis Safe 扩展)
 */
const SAFE_MODULE_ABI = [
  "function safe() view returns (address)",
  "function signer() view returns (address)",
  "function nonce() view returns (uint256)",
  "function frozen() view returns (bool)",
  "function setSigner(address)",
  "function freeze()",
  "function unfreeze()",
  "function delegatedPayEth(address,uint256,uint256,bytes)",
  "function delegatedPayERC20(address,address,uint256,uint256,bytes)",
  "function delegatedExecute(address,uint256,bytes,uint256,bytes)"
];

/**
 * AIEP 类：处理已部署 Agent 合约的交互
 */
class AIEP {
  /**
   * @param {ethers.Provider | ethers.Signer} providerOrSigner Ethers 提供者或签名者
   * @param {string} address Agent 合约地址
   */
  constructor(providerOrSigner, address) {
    this.provider = providerOrSigner;
    this.address = address;
    this.contract = new ethers.Contract(address, ABI, providerOrSigner);
  }

  // --- 基础状态查询 ---
  async ownerOf() { return await this.contract.ownerOf(); }
  async getNonce() { return await this.contract.getNonce(); }
  async balanceOf() { return await this.contract.balanceOf(); }
  async balanceOfERC20(token) { return await this.contract.balanceOfERC20(token); }
  async did() { return await this.contract.did(); }
  async isFrozen() { return await this.contract.isFrozen(); }
  async getAgentSigner() { return await this.contract.getAgentSigner(); }

  // --- 所有者权限操作 ---
  async setAgentSigner(signer) { return await this.contract.setAgentSigner(signer); }
  async transferAgentOwnership(to) { return await this.contract.transferAgentOwnership(to); }
  async updateMetadata(uri) { return await this.contract.updateMetadata(uri); }
  async depositToAgent(amountWei) { return await this.contract.depositToAgent({ value: amountWei }); }
  async depositERC20(token, amount) { return await this.contract.depositERC20(token, amount); }

  // --- 委托支付与执行 (EIP-712) ---

  /**
   * 委托支付原生 ETH
   * @param {string} to 接收者
   * @param {bigint} amountWei 金额
   * @param {number} deadline 截止时间戳
   * @param {ethers.Signer} signer 具备签名权限的热密钥
   */
  async delegatedPayEth(to, amountWei, deadline, signer) {
    const { domain, value, signature } = await this.signDelegatedPayEth(to, amountWei, deadline, signer);
    return await this.contract.delegatedPayEth(to, amountWei, deadline, signature);
  }

  async signDelegatedPayEth(to, amountWei, deadline, signer) {
    const network = await (signer.provider || this.provider).getNetwork();
    const chainId = network.chainId;
    const domain = getDomain(chainId, this.address);
    const nonce = await this.getNonce();
    const value = { to, amount: amountWei, nonce, deadline };
    const signature = await signer.signTypedData(domain, { PayEth: Types.PayEth }, value);
    return { domain, value, signature };
  }

  /**
   * 委托支付 ERC20 代币
   */
  async delegatedPayERC20(token, to, amount, deadline, signer) {
    const { signature } = await this.signDelegatedPayERC20(token, to, amount, deadline, signer);
    return await this.contract.delegatedPayERC20(token, to, amount, deadline, signature);
  }

  async signDelegatedPayERC20(token, to, amount, deadline, signer) {
    const network = await (signer.provider || this.provider).getNetwork();
    const chainId = network.chainId;
    const domain = getDomain(chainId, this.address);
    const nonce = await this.getNonce();
    const value = { token, to, amount, nonce, deadline };
    const signature = await signer.signTypedData(domain, { PayERC20: Types.PayERC20 }, value);
    return { domain, value, signature };
  }

  /**
   * 委托执行任意合约调用
   */
  async delegatedExecute(target, valueWei, data, deadline, signer) {
    const { signature } = await this.signDelegatedExecute(target, valueWei, data, deadline, signer);
    return await this.contract.delegatedExecute(target, valueWei, data, deadline, signature);
  }

  async signDelegatedExecute(target, valueWei, data, deadline, signer) {
    const network = await (signer.provider || this.provider).getNetwork();
    const chainId = network.chainId;
    const domain = getDomain(chainId, this.address);
    const nonce = await this.getNonce();
    const dataHash = ethers.keccak256(data);
    const value = { target, value: valueWei, dataHash, nonce, deadline };
    const signature = await signer.signTypedData(domain, { Execute: Types.Execute }, value);
    return { domain, value, signature };
  }
}

/**
 * AIEPFactory 类：处理 Agent 的创建和预计算地址
 */
class AIEPFactory {
  constructor(providerOrSigner, address) {
    this.provider = providerOrSigner;
    this.address = address;
    this.contract = new ethers.Contract(address, FACTORY_ABI, providerOrSigner);
  }

  async computeAddress(owner, signer, metadataURI, salt) {
    return await this.contract.computeAddress(owner, signer, metadataURI, salt);
  }

  async deployAgent(owner, signer, metadataURI, salt) {
    return await this.contract.deployAgent(owner, signer, metadataURI, salt);
  }
}

/**
 * EasyAgent 类：高级封装，支持“即用即部署”的反事实逻辑
 */
class EasyAgent {
  /**
   * @param {ethers.Provider} provider 
   * @param {string} factoryAddress 
   * @param {ethers.Signer} ownerSigner 所有者签名者 (冷钱包)
   * @param {ethers.Signer} agentSigner 代理签名者 (热密钥)
   * @param {object} opts 配置项 (metadataURI, salt)
   */
  constructor(provider, factoryAddress, ownerSigner, agentSigner, opts = {}) {
    this.provider = provider;
    this.factory = new AIEPFactory(ownerSigner, factoryAddress);
    this.ownerSigner = ownerSigner;
    this.agentSigner = agentSigner;
    this.metadataURI = opts.metadataURI || "ipfs://aetheria-agent-profile";
    this.customSalt = opts.salt;
    this.addressPromise = null;
  }

  /**
   * 预计算 Agent 的合约地址
   */
  async getAddress() {
    if (this.addressPromise) return await this.addressPromise;
    const ownerAddr = await this.ownerSigner.getAddress();
    const agentAddr = await this.agentSigner.getAddress();
    // 如果没有提供自定义 salt，则根据所有者和签名者地址生成确定性 salt
    this.salt = this.customSalt || ethers.keccak256(ethers.toUtf8Bytes(`${ownerAddr}:${agentAddr}`));
    this.owner = ownerAddr;
    this.addressPromise = this.factory.computeAddress(ownerAddr, agentAddr, this.metadataURI, this.salt);
    return await this.addressPromise;
  }

  /**
   * 确保 Agent 已部署在链上
   */
  async ensureDeployed() {
    const addr = await this.getAddress();
    const code = await this.provider.getCode(addr);
    if (code && code !== "0x") return addr;
    const agentAddr = await this.agentSigner.getAddress();
    const tx = await this.factory.deployAgent(this.owner, agentAddr, this.metadataURI, this.salt);
    await tx.wait();
    return addr;
  }

  /**
   * 支付 ERC20 代币 (如果未部署，则合并部署与支付交易)
   */
  async payERC20(token, to, amount, deadlineSec, opts = {}) {
    const addr = await this.getAddress();
    const deadline = deadlineSec || Math.floor(Date.now() / 1000) + 3600;
    const agent = new AIEP(this.ownerSigner, addr);
    const normalizedAmount = await normalizeAmountERC20(this.provider, token, amount);
    const code = await this.provider.getCode(addr);
    const toAddr = await resolveToAddress(this.provider, to);

    // 自动补足资金逻辑
    if (opts.autoRefill) {
      const min = await normalizeAmountERC20(this.provider, token, opts.autoRefill.minToken || amount);
      const bal = (code && code !== "0x") ? await agent.balanceOfERC20(token) : 0n;
      if (bal < min) {
        const refillAmt = opts.autoRefill.refillAmount || (min - bal);
        await this.fundERC20(token, refillAmt);
      }
    }

    if (!code || code === "0x") {
      // 合并部署交易
      const chainId = (await this.provider.getNetwork()).chainId;
      const domain = getDomain(chainId, addr);
      const value = { token, to: toAddr, amount: normalizedAmount, nonce: 0, deadline };
      const signature = await this.agentSigner.signTypedData(domain, { PayERC20: Types.PayERC20 }, value);
      const agentAddr = await this.agentSigner.getAddress();
      return await this.factory.contract.deployAndDelegatedPayERC20(
        this.owner, agentAddr, this.metadataURI, this.salt, 
        token, toAddr, normalizedAmount, deadline, signature
      );
    }
    return await agent.delegatedPayERC20(token, toAddr, normalizedAmount, deadline, this.agentSigner);
  }

  /**
   * 支付 ETH (如果未部署，则合并部署与支付交易)
   */
  async payEth(to, amountWei, deadlineSec, opts = {}) {
    const addr = await this.getAddress();
    const deadline = deadlineSec || Math.floor(Date.now() / 1000) + 3600;
    const agent = new AIEP(this.ownerSigner, addr);
    const code = await this.provider.getCode(addr);
    const toAddr = await resolveToAddress(this.provider, to);
    const wei = await normalizeWei(amountWei);

    if (opts.autoRefill) {
      const min = await normalizeWei(opts.autoRefill.minWei || wei);
      const bal = (code && code !== "0x") ? await agent.balanceOf() : 0n;
      if (bal < min) {
        const refillWei = await normalizeWei(opts.autoRefill.refillWei || (min - bal));
        await this.fundEth(refillWei);
      }
    }

    if (!code || code === "0x") {
      const chainId = (await this.provider.getNetwork()).chainId;
      const domain = getDomain(chainId, addr);
      const value = { to: toAddr, amount: wei, nonce: 0, deadline };
      const signature = await this.agentSigner.signTypedData(domain, { PayEth: Types.PayEth }, value);
      const agentAddr = await this.agentSigner.getAddress();
      return await this.factory.contract.deployAndDelegatedPayEth(
        this.owner, agentAddr, this.metadataURI, this.salt, 
        toAddr, wei, deadline, signature
      );
    }
    return await agent.delegatedPayEth(toAddr, wei, deadline, this.agentSigner);
  }

  /**
   * 给 Agent 充值 ETH
   */
  async fundEth(amount) {
    const addr = await this.getAddress();
    const wei = await normalizeWei(amount);
    const code = await this.provider.getCode(addr);
    if (code && code !== '0x') {
      const agent = new AIEP(this.ownerSigner, addr);
      return await agent.depositToAgent(wei);
    }
    return await this.ownerSigner.sendTransaction({ to: addr, value: wei });
  }

  /**
   * 给 Agent 充值 ERC20
   */
  async fundERC20(token, amount) {
    const addr = await this.getAddress();
    const amt = await normalizeAmountERC20(this.provider, token, amount);
    const erc20 = new ethers.Contract(token, ["function transfer(address,uint256) returns (bool)"], this.ownerSigner);
    return await erc20.transfer(addr, amt);
  }

  async getStatus(token = null) {
    const addr = await this.getAddress();
    const code = await this.provider.getCode(addr);
    const deployed = !!(code && code !== '0x');
    let ethBalance = 0n;
    let erc20Balance = null;
    if (deployed) {
      const agent = new AIEP(this.provider, addr);
      ethBalance = await agent.balanceOf();
      if (token) erc20Balance = await agent.balanceOfERC20(token);
    }
    return { address: addr, deployed, ethBalance, erc20Balance };
  }
}

/**
 * 辅助函数：格式化 ERC20 金额 (支持字符串形式的单位，如 "1.5")
 */
async function normalizeAmountERC20(provider, token, amount) {
  if (typeof amount === 'bigint') return amount;
  if (typeof amount === 'number') return BigInt(amount);
  if (typeof amount === 'string') {
    const erc20 = new ethers.Contract(token, ["function decimals() view returns (uint8)"], provider);
    const decimals = await erc20.decimals();
    return ethers.parseUnits(amount, decimals);
  }
  throw new Error("Invalid amount type: must be string, number or bigint");
}

/**
 * 辅助函数：格式化 ETH 金额
 */
async function normalizeWei(amount) {
  if (typeof amount === 'bigint') return amount;
  if (typeof amount === 'number') return BigInt(amount);
  if (typeof amount === 'string') return ethers.parseEther(amount);
  throw new Error('Invalid wei amount: must be string, number or bigint');
}

/**
 * 辅助函数：解析地址 (支持 ENS，如 "vitalik.eth")
 */
async function resolveToAddress(provider, to) {
  if (typeof to !== 'string') return to;
  if (to.endsWith('.eth')) {
    const resolved = await provider.resolveName(to);
    if (!resolved) throw new Error(`ENS name could not be resolved: ${to}`);
    return resolved;
  }
  return to;
}

/**
 * 导出模块
 */
module.exports = {
  AIEP,
  AIEPFactory,
  EasyAgent,
  Types,
  getDomain,
  buildDid: (chainId, contract) => `did:ethr:${chainId}:${contract.toLowerCase()}`
};

File: sdk/python/aiep.py
"""
AIEP (AI-Agent Identity & Execution Protocol) Python SDK
用于与 AetheriaAgentDID 智能合约交互的官方 Python SDK。
支持 DID 身份管理、EIP-712 委托支付和执行、以及反事实部署逻辑。
"""

import json
import time
from typing import Any, Dict, List, Optional
from decimal import Decimal
from web3 import Web3
from eth_account import Account
from eth_account.messages import encode_structured_data
from eth_utils import keccak, to_checksum_address

# --- 合约 ABI 定义 ---

ABI = [
    {"type":"function","name":"ownerOf","stateMutability":"view","inputs":[],"outputs":[{"name":"","type":"address"}]},
    {"type":"function","name":"setAgentSigner","stateMutability":"nonpayable","inputs":[{"name":"signer","type":"address"}],"outputs":[]},
    {"type":"function","name":"transferAgentOwnership","stateMutability":"nonpayable","inputs":[{"name":"newOwner","type":"address"}],"outputs":[]},
    {"type":"function","name":"getNonce","stateMutability":"view","inputs":[],"outputs":[{"name":"","type":"uint256"}]},
    {"type":"function","name":"depositToAgent","stateMutability":"payable","inputs":[],"outputs":[]},
    {"type":"function","name":"balanceOf","stateMutability":"view","inputs":[],"outputs":[{"name":"","type":"uint256"}]},
    {"type":"function","name":"delegatedPayEth","stateMutability":"nonpayable","inputs":[{"name":"to","type":"address"},{"name":"amount","type":"uint256"},{"name":"deadline","type":"uint256"},{"name":"signature","type":"bytes"}],"outputs":[]},
    {"type":"function","name":"depositERC20","stateMutability":"nonpayable","inputs":[{"name":"token","type":"address"},{"name":"amount","type":"uint256"}],"outputs":[]},
    {"type":"function","name":"balanceOfERC20","stateMutability":"view","inputs":[{"name":"token","type":"address"}],"outputs":[{"name":"","type":"uint256"}]},
    {"type":"function","name":"delegatedPayERC20","stateMutability":"nonpayable","inputs":[{"name":"token","type":"address"},{"name":"to","type":"address"},{"name":"amount","type":"uint256"},{"name":"deadline","type":"uint256"},{"name":"signature","type":"bytes"}],"outputs":[]},
    {"type":"function","name":"updateMetadata","stateMutability":"nonpayable","inputs":[{"name":"metadataURI","type":"string"}],"outputs":[]},
    {"type":"function","name":"did","stateMutability":"view","inputs":[],"outputs":[{"name":"","type":"string"}]},
    {"type":"function","name":"isFrozen","stateMutability":"view","inputs":[],"outputs":[{"name":"","type":"bool"}]},
    {"type":"function","name":"getAgentSigner","stateMutability":"view","inputs":[],"outputs":[{"name":"","type":"address"}]},
    {"type":"function","name":"delegatedExecute","stateMutability":"nonpayable","inputs":[{"name":"target","type":"address"},{"name":"value","type":"uint256"},{"name":"data","type":"bytes"},{"name":"deadline","type":"uint256"},{"name":"signature","type":"bytes"}],"outputs":[]}
]

FACTORY_ABI = [
    {"type":"function","name":"deployAgent","stateMutability":"nonpayable","inputs":[
        {"name":"owner","type":"address"}, {"name":"signer","type":"address"}, {"name":"metadataURI","type":"string"}, {"name":"salt","type":"bytes32"}
    ],"outputs":[{"name":"","type":"address"}]},
    {"type":"function","name":"computeAddress","stateMutability":"view","inputs":[
        {"name":"owner","type":"address"}, {"name":"signer","type":"address"}, {"name":"metadataURI","type":"string"}, {"name":"salt","type":"bytes32"}
    ],"outputs":[{"name":"","type":"address"}]},
    {"type":"function","name":"deployAndDelegatedPayERC20","stateMutability":"nonpayable","inputs":[
        {"name":"owner","type":"address"}, {"name":"signer","type":"address"}, {"name":"metadataURI","type":"string"}, {"name":"salt","type":"bytes32"},
        {"name":"token","type":"address"}, {"name":"to","type":"address"}, {"name":"amount","type":"uint256"}, {"name":"deadline","type":"uint256"}, {"name":"signature","type":"bytes"}
    ],"outputs":[{"name":"","type":"address"}]},
    {"type":"function","name":"deployAndDelegatedPayEth","stateMutability":"nonpayable","inputs":[
        {"name":"owner","type":"address"}, {"name":"signer","type":"address"}, {"name":"metadataURI","type":"string"}, {"name":"salt","type":"bytes32"},
        {"name":"to","type":"address"}, {"name":"amount","type":"uint256"}, {"name":"deadline","type":"uint256"}, {"name":"signature","type":"bytes"}
    ],"outputs":[{"name":"","type":"address"}]},
    {"type":"function","name":"deployAndDelegatedExecute","stateMutability":"nonpayable","inputs":[
        {"name":"owner","type":"address"}, {"name":"signer","type":"address"}, {"name":"metadataURI","type":"string"}, {"name":"salt","type":"bytes32"},
        {"name":"target","type":"address"}, {"name":"value","type":"uint256"}, {"name":"data","type":"bytes"}, {"name":"deadline","type":"uint256"}, {"name":"signature","type":"bytes"}
    ],"outputs":[{"name":"","type":"address"}]}
]

# --- EIP-712 配置 ---

TYPES = {
    "EIP712Domain": [
        {"name": "name", "type": "string"},
        {"name": "version", "type": "string"},
        {"name": "chainId", "type": "uint256"},
        {"name": "verifyingContract", "type": "address"}
    ],
    "PayEth": [
        {"name": "to", "type": "address"},
        {"name": "amount", "type": "uint256"},
        {"name": "nonce", "type": "uint256"},
        {"name": "deadline", "type": "uint256"}
    ],
    "PayERC20": [
        {"name": "token", "type": "address"},
        {"name": "to", "type": "address"},
        {"name": "amount", "type": "uint256"},
        {"name": "nonce", "type": "uint256"},
        {"name": "deadline", "type": "uint256"}
    ],
    "Execute": [
        {"name": "target", "type": "address"},
        {"name": "value", "type": "uint256"},
        {"name": "dataHash", "type": "bytes32"},
        {"name": "nonce", "type": "uint256"},
        {"name": "deadline", "type": "uint256"}
    ]
}

def get_domain(chain_id: int, verifying_contract: str) -> Dict[str, Any]:
    return {
        "name": "AetheriaAgentDID",
        "version": "1",
        "chainId": chain_id,
        "verifyingContract": to_checksum_address(verifying_contract)
    }

# --- 核心类实现 ---

class AIEP:
    """处理已部署 Agent 合约的交互"""
    def __init__(self, w3: Web3, contract_address: str):
        self.w3 = w3
        self.address = to_checksum_address(contract_address)
        self.contract = w3.eth.contract(address=self.address, abi=ABI)

    def _get_tx_params(self, from_address: str, value: int = 0) -> Dict[str, Any]:
        return {
            'from': from_address,
            'nonce': self.w3.eth.get_transaction_count(from_address),
            'value': value,
            'gasPrice': self.w3.eth.gas_price,
            'chainId': self.w3.eth.chain_id
        }

    def _send_tx(self, func, private_key: str, value: int = 0):
        acct = Account.from_key(private_key)
        params = self._get_tx_params(acct.address, value)
        tx = func.build_transaction(params)
        tx['gas'] = self.w3.eth.estimate_gas(tx)
        signed = Account.sign_transaction(tx, private_key)
        tx_hash = self.w3.eth.send_raw_transaction(signed.raw_transaction)
        return self.w3.eth.wait_for_transaction_receipt(tx_hash)

    # 状态查询
    def owner_of(self) -> str: return self.contract.functions.ownerOf().call()
    def get_nonce(self) -> int: return self.contract.functions.getNonce().call()
    def balance_of(self) -> int: return self.contract.functions.balanceOf().call()
    def balance_of_erc20(self, token: str) -> int: return self.contract.functions.balanceOfERC20(to_checksum_address(token)).call()
    def did(self) -> str: return self.contract.functions.did().call()

    # 委托操作签名
    def _sign_typed_data(self, primary_type: str, message: Dict[str, Any], private_key: str) -> bytes:
        structured_data = {
            "types": TYPES,
            "domain": get_domain(self.w3.eth.chain_id, self.address),
            "primaryType": primary_type,
            "message": message
        }
        signed = Account.sign_message(encode_structured_data(structured_data), private_key)
        return signed.signature

    def delegated_pay_eth(self, to: str, amount_wei: int, deadline: int, signer_private_key: str, relayer_private_key: str):
        nonce = self.get_nonce()
        message = {"to": to_checksum_address(to), "amount": amount_wei, "nonce": nonce, "deadline": deadline}
        signature = self._sign_typed_data("PayEth", message, signer_private_key)
        func = self.contract.functions.delegatedPayEth(to_checksum_address(to), amount_wei, deadline, signature)
        return self._send_tx(func, relayer_private_key)

    def delegated_execute(self, target: str, value_wei: int, data: bytes, deadline: int, signer_private_key: str, relayer_private_key: str):
        nonce = self.get_nonce()
        data_hash = keccak(data)
        message = {"target": to_checksum_address(target), "value": value_wei, "dataHash": data_hash, "nonce": nonce, "deadline": deadline}
        signature = self._sign_typed_data("Execute", message, signer_private_key)
        func = self.contract.functions.delegatedExecute(to_checksum_address(target), value_wei, data, deadline, signature)
        return self._send_tx(func, relayer_private_key)

class EasyAgent:
    """高级封装：支持反事实部署与自动资金管理"""
    def __init__(self, w3: Web3, factory_address: str, owner_private_key: str, agent_signer_private_key: str, metadata_uri: str = "ipfs://agent"):
        self.w3 = w3
        self.factory_address = to_checksum_address(factory_address)
        self.factory = w3.eth.contract(address=self.factory_address, abi=FACTORY_ABI)
        self.owner_acct = Account.from_key(owner_private_key)
        self.owner_pk = owner_private_key
        self.agent_acct = Account.from_key(agent_signer_private_key)
        self.agent_pk = agent_signer_private_key
        self.metadata_uri = metadata_uri
        
        # 确定性 Salt 生成
        salt_src = f"{self.owner_acct.address}:{self.agent_acct.address}".encode()
        self.salt = keccak(salt_src)
        self.address = self.factory.functions.computeAddress(
            self.owner_acct.address, self.agent_acct.address, self.metadata_uri, self.salt
        ).call()

    def is_deployed(self) -> bool:
        code = self.w3.eth.get_code(self.address)
        return len(code) > 0

    def ensure_deployed(self):
        if self.is_deployed(): return self.address
        func = self.factory.functions.deployAgent(
            self.owner_acct.address, self.agent_acct.address, self.metadata_uri, self.salt
        )
        self._send_factory_tx(func)
        return self.address

    def _send_factory_tx(self, func):
        params = {
            'from': self.owner_acct.address,
            'nonce': self.w3.eth.get_transaction_count(self.owner_acct.address),
            'gasPrice': self.w3.eth.gas_price,
            'chainId': self.w3.eth.chain_id
        }
        tx = func.build_transaction(params)
        tx['gas'] = self.w3.eth.estimate_gas(tx)
        signed = Account.sign_transaction(tx, self.owner_pk)
        tx_hash = self.w3.eth.send_raw_transaction(signed.raw_transaction)
        return self.w3.eth.wait_for_transaction_receipt(tx_hash)

    def pay_eth(self, to: str, amount_wei: int, deadline: Optional[int] = None):
        """自动处理部署并支付 ETH"""
        dl = deadline or int(time.time()) + 3600
        if not self.is_deployed():
            # 反事实部署并执行
            domain = get_domain(self.w3.eth.chain_id, self.address)
            message = {"to": to_checksum_address(to), "amount": amount_wei, "nonce": 0, "deadline": dl}
            structured_data = {"types": TYPES, "domain": domain, "primaryType": "PayEth", "message": message}
            sig = Account.sign_message(encode_structured_data(structured_data), self.agent_pk).signature
            
            func = self.factory.functions.deployAndDelegatedPayEth(
                self.owner_acct.address, self.agent_acct.address, self.metadata_uri, self.salt,
                to_checksum_address(to), amount_wei, dl, sig
            )
            return self._send_factory_tx(func)
        
        return AIEP(self.w3, self.address).delegated_pay_eth(to, amount_wei, dl, self.agent_pk, self.owner_pk)

    def fund_eth(self, amount_wei: int):
        """为 Agent 充值 ETH"""
        tx = {
            'from': self.owner_acct.address,
            'to': self.address,
            'value': amount_wei,
            'nonce': self.w3.eth.get_transaction_count(self.owner_acct.address),
            'gasPrice': self.w3.eth.gas_price,
            'chainId': self.w3.eth.chain_id
        }
        tx['gas'] = self.w3.eth.estimate_gas(tx)
        signed = Account.sign_transaction(tx, self.owner_pk)
        return self.w3.eth.wait_for_transaction_receipt(self.w3.eth.send_raw_transaction(signed.raw_transaction))

    def _normalize_erc20_amount(self, token: str, amount: Any) -> int:
        if isinstance(amount, (int, float, Decimal)):
            # 如果是数字，假设需要根据精度转换
            erc20 = self.w3.eth.contract(address=to_checksum_address(token), abi=[
                {"type":"function","name":"decimals","inputs":[],"outputs":[{"type":"uint8"}]}
            ])
            decimals = erc20.functions.decimals().call()
            return int(Decimal(str(amount)) * (Decimal(10) ** decimals))
        return int(amount)

# --- 工具函数 ---

def build_did(chain_id: int, address: str) -> str:
    """生成符合规范的 DID"""
    return f"did:ethr:{chain_id}:{to_checksum_address(address).lower()}"
